### ISSUE TITLE
[Bug] Vits with weighted sampler sees training time increase with each epoch

### ISSUE THREAD

**Issue by MarkuzK** [coqui-ai/TTS#2560]
### Describe the bug

Training a VITS model from scratch with the following deviations from the recipe:
- sample_rate         = 22050,
- use_speaker_embedding   = False,
- use_d_vector_file       = True,
- d_vector_file           = speaker_embeddings_list,    # pre-computed
- use_weighted_sampler = True,
- weighted_sampler_attrs = {"speaker_name": 1.0},
- weighted_sampler_multipliers = {}

Data-set is custom; it includes roughly 4,000 speakers with 1,000,000 recordings.

Using a batch size of 24, 1 epoch takes just shy of 40k steps. Training speed is significantly impacted by the weighted sampler; some time measurements (taken every 5k steps):
-     5,000 - 10,000: 43 minutes
-     10,000 - 15,000: 44 minutes
-     15,000 - 20,000: 46 minutes
-     20,000 - 25,000: 47 minutes
-     25,000 - 30,000: 49 minutes
-     30,000 - 35,000: 50 minutes
-     35,000 - 40,000: 58 minutes (*)
-     40,000 - 45,000: 54 minutes
-     45,000 - 50,000: 56 minutes
-     50,000 - 55,000: 57 minutes
-     55,000 - 60,000: 59 minutes
-     60,000 - 65,000: 61 minutes
-     65,000 - 70,000: 61 minutes
-     70,000 - 75,000: 64 minutes
-     75,000 - 80,000: 70 minutes (*)
-     80,000 - 85,000: 69 minutes
-     85,000 - 90,000: 69 minutes
-     90,000 - 95,000: 72 minutes
-     95,000 - 100,000: 75 minutes
-     100,000 - 105,000: 75 minutes
-     105,000 - 110,000: 77 minutes
-     110,000 - 115,000: 80 minutes
-     115,000 - 120,000: 86 minutes (*)
-     120,000 - 125,000: 84 minutes

(*) ... indicated that eval took place and processing time increase is expected.

When I uncomment the following 3 lines, processing speed remains consistent:
- #use_weighted_sampler = True,
- #weighted_sampler_attrs = {"speaker_name": 1.0},
- #weighted_sampler_multipliers = {}

In that case (and with identical setup), each step takes about 38-39 minutes (confirmed until 120k steps).

### To Reproduce

config.json with dataset references removed

{
    "output_path": "results/baseline-models",
    "logger_uri": null,
    "run_name": "vits",
    "project_name": null,
    "run_description": "\ud83d\udc38Coqui trainer run.",
    "print_step": 50,
    "plot_step": 100,
    "model_param_stats": false,
    "wandb_entity": null,
    "dashboard_logger": "tensorboard",
    "log_model_step": 5000,
    "save_step": 5000,
    "save_n_checkpoints": 100,
    "save_checkpoints": true,
    "save_all_best": true,
    "save_best_after": 10000,
    "target_loss": null,
    "print_eval": true,
    "test_delay_epochs": -1,
    "run_eval": true,
    "run_eval_steps": null,
    "distributed_backend": "nccl",
    "distributed_url": "tcp://localhost:54321",
    "mixed_precision": true,
    "epochs": 100,
    "batch_size": 24,
    "eval_batch_size": 16,
    "grad_clip": [
        1000,
        1000
    ],
    "scheduler_after_epoch": true,
    "lr": 0.001,
    "optimizer": "AdamW",
    "optimizer_params": {
        "betas": [
            0.8,
            0.99
        ],
        "eps": 1e-09,
        "weight_decay": 0.01
    },
    "lr_scheduler": null,
    "lr_scheduler_params": {},
    "use_grad_scaler": false,
    "cudnn_enable": true,
    "cudnn_deterministic": false,
    "cudnn_benchmark": false,
    "training_seed": 54321,
    "model": "vits",
    "num_loader_workers": 4,
    "num_eval_loader_workers": 4,
    "use_noise_augment": false,
    "audio": {
        "fft_size": 1024,
        "sample_rate": 22050,
        "win_length": 1024,
        "hop_length": 256,
        "num_mels": 80,
        "mel_fmin": 0,
        "mel_fmax": null
    },
    "use_phonemes": false,
    "phonemizer": null,
    "phoneme_language": "en-us",
    "compute_input_seq_cache": true,
    "text_cleaner": "english_cleaners",
    "enable_eos_bos_chars": false,
    "test_sentences_file": "",
    "phoneme_cache_path": "results/baseline-models/phoneme_cache",
    "characters": null,
    "add_blank": true,
    "batch_group_size": 0,
    "loss_masking": null,
    "min_audio_len": 1,
    "max_audio_len": Infinity,
    "min_text_len": 1,
    "max_text_len": 325,
    "compute_f0": false,
    "compute_energy": false,
    "compute_linear_spec": true,
    "precompute_num_workers": 0,
    "start_by_longest": false,
    "shuffle": false,
    "drop_last": false,
    "datasets": [
        {
            "formatter": "vctk",
            "dataset_name": "VCTK",
            "path": "results/datasets/<CUSTOM>",
            "meta_file_train": "",
            "ignored_speakers": null,
            "language": "en-us",
            "phonemizer": "",
            "meta_file_val": "",
            "meta_file_attn_mask": ""
        },
    ],
    "test_sentences": [
        [
            "It took me quite a long time to develop a voice, and now that I have it I'm not going to be silent.",
            "VCTK_custom"
        ],
    ],
    "eval_split_max_size": null,
    "eval_split_size": 0.01,
    "use_speaker_weighted_sampler": false,
    "speaker_weighted_sampler_alpha": 1.0,
    "use_language_weighted_sampler": false,
    "language_weighted_sampler_alpha": 1.0,
    "use_length_weighted_sampler": false,
    "length_weighted_sampler_alpha": 1.0,
    "model_args": {
        "num_chars": 67,
        "out_channels": 513,
        "spec_segment_size": 32,
        "hidden_channels": 192,
        "hidden_channels_ffn_text_encoder": 768,
        "num_heads_text_encoder": 2,
        "num_layers_text_encoder": 10,
        "kernel_size_text_encoder": 3,
        "dropout_p_text_encoder": 0.1,
        "dropout_p_duration_predictor": 0.5,
        "kernel_size_posterior_encoder": 5,
        "dilation_rate_posterior_encoder": 1,
        "num_layers_posterior_encoder": 16,
        "kernel_size_flow": 5,
        "dilation_rate_flow": 1,
        "num_layers_flow": 4,
        "resblock_type_decoder": "1",
        "resblock_kernel_sizes_decoder": [
            3,
            7,
            11
        ],
        "resblock_dilation_sizes_decoder": [
            [
                1,
                3,
                5
            ],
            [
                1,
                3,
                5
            ],
            [
                1,
                3,
                5
            ]
        ],
        "upsample_rates_decoder": [
            8,
            8,
            2,
            2
        ],
        "upsample_initial_channel_decoder": 512,
        "upsample_kernel_sizes_decoder": [
            16,
            16,
            4,
            4
        ],
        "periods_multi_period_discriminator": [
            2,
            3,
            5,
            7,
            11
        ],
        "use_sdp": true,
        "noise_scale": 1.0,
        "inference_noise_scale": 0.667,
        "length_scale": 1,
        "noise_scale_dp": 1.0,
        "inference_noise_scale_dp": 1.0,
        "max_inference_len": null,
        "init_discriminator": true,
        "use_spectral_norm_disriminator": false,
        "use_speaker_embedding": false,
        "num_speakers": 0,
        "speakers_file": "results/baseline-models/vits_April-25-2023_04+12PM-ab5edc8/speakers.pth",
        "d_vector_file": [
            "results/datasets/<CUSTOM>/speakers.pth",
        ],
        "speaker_embedding_channels": 256,
        "use_d_vector_file": true,
        "d_vector_dim": 512,
        "detach_dp_input": true,
        "use_language_embedding": false,
        "embedded_language_dim": 4,
        "num_languages": 0,
        "language_ids_file": null,
        "use_speaker_encoder_as_loss": false,
        "speaker_encoder_config_path": "",
        "speaker_encoder_model_path": "",
        "condition_dp_on_speaker": true,
        "freeze_encoder": false,
        "freeze_DP": false,
        "freeze_PE": false,
        "freeze_flow_decoder": false,
        "freeze_waveform_decoder": false,
        "encoder_sample_rate": null,
        "interpolate_z": true,
        "reinit_DP": false,
        "reinit_text_encoder": false
    },
    "lr_gen": 0.0002,
    "lr_disc": 0.0002,
    "lr_scheduler_gen": "ExponentialLR",
    "lr_scheduler_gen_params": {
        "gamma": 0.999875,
        "last_epoch": -1
    },
    "lr_scheduler_disc": "ExponentialLR",
    "lr_scheduler_disc_params": {
        "gamma": 0.999875,
        "last_epoch": -1
    },
    "kl_loss_alpha": 1.0,
    "disc_loss_alpha": 1.0,
    "gen_loss_alpha": 1.0,
    "feat_loss_alpha": 1.0,
    "mel_loss_alpha": 45.0,
    "dur_loss_alpha": 1.0,
    "speaker_encoder_loss_alpha": 1.0,
    "return_wav": true,
    "use_weighted_sampler": true,
    "weighted_sampler_attrs": {
        "speaker_name": 1.0
    },
    "weighted_sampler_multipliers": {},
    "r": 1,
    "num_speakers": 0,
    "use_speaker_embedding": false,
    "speakers_file": "results/baseline-models/vits_April-25-2023_04+12PM-ab5edc8/speakers.pth",
    "speaker_embedding_channels": 256,
    "language_ids_file": null,
    "use_language_embedding": false,
    "use_d_vector_file": true,
    "d_vector_file": [
        "results/datasets/<CUSTOM>/speakers.pth",
    ],
    "d_vector_dim": 512
}

### Expected behavior

Somewhat consistent training times.

### Logs

```shell
> Training Environment:
 | > Current device: 0
 | > Num. of GPUs: 1
 | > Num. of CPUs: 64
 | > Num. of Torch Threads: 32
 | > Torch seed: 54321
 | > Torch CUDNN: True
 | > Torch CUDNN deterministic: False
 | > Torch CUDNN benchmark: False
 > Start Tensorboard: tensorboard --logdir=results/baseline-models/vits_April-25-2023_04+12PM-ab5edc8

 > Model has 93869740 parameters

 > EPOCH: 0/100
 --> results/baseline-models/vits_April-25-2023_04+12PM-ab5edc8

 > TRAINING (2023-04-25 16:15:23) 

   --> STEP: 0/39978 -- GLOBAL_STEP: 0
     | > loss_disc: 6.08416  (6.08416)
     | > loss_disc_real_0: 1.00759  (1.00759)
     | > loss_disc_real_1: 0.99431  (0.99431)
     | > loss_disc_real_2: 1.00378  (1.00378)
     | > loss_disc_real_3: 1.03363  (1.03363)
     | > loss_disc_real_4: 1.03340  (1.03340)
     | > loss_disc_real_5: 1.01084  (1.01084)
     | > loss_0: 6.08416  (6.08416)
     | > grad_norm_0: 0.00000  (0.00000)
     | > loss_gen: 6.08405  (6.08405)
     | > loss_kl: 183.77365  (183.77365)
     | > loss_feat: 0.20418  (0.20418)
     | > loss_mel: 87.52084  (87.52084)
     | > loss_duration: 1.70976  (1.70976)
     | > amp_scaler: 32768.00000  (32768.00000)
     | > loss_1: 279.29245  (279.29245)
     | > grad_norm_1: 0.00000  (0.00000)
     | > current_lr_0: 0.00020 
     | > current_lr_1: 0.00020 
     | > step_time: 2.71020  (2.71022)
     | > loader_time: 3.03020  (3.03017)


   --> STEP: 50/39978 -- GLOBAL_STEP: 50
     | > loss_disc: 2.68518  (3.07987)
     | > loss_disc_real_0: 0.19443  (0.28438)
     | > loss_disc_real_1: 0.28188  (0.29706)
     | > loss_disc_real_2: 0.27416  (0.29688)
     | > loss_disc_real_3: 0.26140  (0.31480)
     | > loss_disc_real_4: 0.30015  (0.30814)
     | > loss_disc_real_5: 0.15024  (0.30518)

[...]

 > CHECKPOINT : results/baseline-models/vits_April-25-2023_04+12PM-ab5edc8/checkpoint_125000.pth

   --> STEP: 5116/39978 -- GLOBAL_STEP: 125050
     | > loss_disc: 2.59933  (2.50169)
     | > loss_disc_real_0: 0.16436  (0.15754)
     | > loss_disc_real_1: 0.21442  (0.21388)
     | > loss_disc_real_2: 0.26732  (0.22891)
     | > loss_disc_real_3: 0.25044  (0.22650)
     | > loss_disc_real_4: 0.24786  (0.23146)
     | > loss_disc_real_5: 0.22129  (0.23358)
     | > loss_0: 2.59933  (2.50169)
     | > grad_norm_0: 8.70634  (14.36523)
     | > loss_gen: 2.31212  (2.26625)
     | > loss_kl: 2.77006  (2.72223)
     | > loss_feat: 5.63128  (5.92124)
     | > loss_mel: 23.32931  (22.20046)
     | > loss_duration: 1.95802  (1.88702)
     | > amp_scaler: 256.00000  (343.36826)
     | > loss_1: 36.00079  (34.99720)
     | > grad_norm_1: 85.70525  (135.23337)
     | > current_lr_0: 0.00020 
     | > current_lr_1: 0.00020 
     | > step_time: 0.49240  (1.00271)
     | > loader_time: 0.00490  (0.00455)


   --> STEP: 5166/39978 -- GLOBAL_STEP: 125100
     | > loss_disc: 2.39309  (2.50171)
     | > loss_disc_real_0: 0.13230  (0.15750)
     | > loss_disc_real_1: 0.24009  (0.21389)
     | > loss_disc_real_2: 0.24266  (0.22893)
     | > loss_disc_real_3: 0.27318  (0.22653)
     | > loss_disc_real_4: 0.23696  (0.23148)
     | > loss_disc_real_5: 0.25497  (0.23360)
     | > loss_0: 2.39309  (2.50171)
     | > grad_norm_0: 13.38440  (14.36817)
     | > loss_gen: 2.40192  (2.26614)
     | > loss_kl: 2.82925  (2.72230)
     | > loss_feat: 5.41125  (5.92106)
     | > loss_mel: 21.54952  (22.20061)
     | > loss_duration: 1.85243  (1.88689)
     | > amp_scaler: 256.00000  (342.52265)
     | > loss_1: 34.04438  (34.99700)
     | > grad_norm_1: 86.93588  (135.23940)
     | > current_lr_0: 0.00020 
     | > current_lr_1: 0.00020 
     | > step_time: 1.16550  (1.00197)
     | > loader_time: 0.00370  (0.00455)
```

### Environment

```shell
{
    "CUDA": {
        "GPU": [
            "NVIDIA GeForce RTX 3090"
        ],
        "available": true,
        "version": "11.7"
    },
    "Packages": {
        "PyTorch_debug": false,
        "PyTorch_version": "1.13.0+cu117",
        "TTS": "0.13.3",
        "numpy": "1.22.4"
    },
    "System": {
        "OS": "Linux",
        "architecture": [
            "64bit",
            "ELF"
        ],
        "processor": "x86_64",
        "python": "3.10.6",
        "version": "#202303130630~1681329778~22.04~d824cd4 SMP PREEMPT_DYNAMIC Wed A"
    }
}
```

**Comment by erogol** [comment]
it is normal. If you don't use sampler, it sorts samples by length so GPUs are well-utilized. With sampler, order is not promised and batches are wider with more padding thus step time increases.
